File: maze.go
// Maze generator in Go
// Joe Wingbermuehle
// 2012-08-07

package main

import (
   "fmt"
   "math/rand"
   "time"
)

const (
   WALL     = 0
   SPACE    = 1
)

type Maze struct {
   width, height int
   data [][]byte
}

/** Create an empty maze.
 * @param w The width (must be odd).
 * @param h The height (must be odd).
 */
func NewMaze(w int, h int) *Maze {
   m := Maze { w, h, make([][]byte, h) }
   for y := range m.data {
      m.data[y] = make([]byte, w)
      for x := range m.data[y] {
         m.data[y][x] = WALL
      }
   }
   for x := 0; x < w; x++ {
      m.data[0][x], m.data[h - 1][x] = SPACE, SPACE
   }
   for y := 0; y < h; y++ {
      m.data[y][0], m.data[y][w - 1] = SPACE, SPACE
   }
   return &m
}

/** Start carving a maze at the specified coordinates. */
func CarveMaze(m *Maze, r *rand.Rand, x int, y int) {
   directions := [][]int { {1, 0}, {-1, 0}, {0, 1}, {0, -1} }
   d := r.Intn(4)
   for i := 0; i < 4; i++ {
      dx, dy := directions[d][0], directions[d][1]
      ax, ay := x + dx, y + dy
      bx, by := ax + dx, ay + dy
      if m.data[ay][ax] == WALL && m.data[by][bx] == WALL {
         m.data[ay][ax], m.data[by][bx] = SPACE, SPACE
         CarveMaze(m, r, bx, by) 
      }
      d = (d + 1) % 4
   }
}

/** Generate a maze. */
func GenerateMaze(m *Maze) {
   r := rand.New(rand.NewSource(time.Now().Unix()))
   m.data[2][2] = SPACE
   CarveMaze(m, r, 2, 2)
   m.data[1][2] = SPACE
   m.data[m.height - 2][m.width - 3] = SPACE
}

/** Show a generated maze. */
func ShowMaze(m *Maze) {
   for y := 0; y < m.height; y++ {
      for x := 0; x < m.width; x++ {
         if m.data[y][x] == WALL {
            fmt.Printf("[]")
         } else {
            fmt.Printf("  ")
         }
      }
      fmt.Printf("\n")
   }
}

func main() {
   m := NewMaze(39, 23)
   GenerateMaze(m)
   ShowMaze(m)
}

File: maze.nim
# Maze generator in Nimrod
# Joe Wingbermuehle 2013-10-01

import math

# Width and height must be odd.
const width = 39
const height = 23

type MazeT = array[0 .. height - 1, array[0 .. width - 1, int]]

proc showMaze(maze: MazeT) =
   for y in countup(0, height - 1):
      for x in countup(0, width - 1):
         if maze[y][x] == 1:
            write(stdout, "[]")
         else:
            write(stdout, "  ")
      write(stdout, "\n")

proc initMaze(maze: ref MazeT) =
   for y in countup(0, height - 1):
      for x in countup(0, width - 1):
         maze[y][x] = 1
   for x in countup(0, width - 1):
      maze[0][x] = 0
      maze[height - 1][x] = 0
   for y in countup(0, height - 1):
      maze[y][0] = 0
      maze[y][width - 1] = 0

proc carveMaze(maze: ref MazeT, x, y: int) =
   maze[y][x] = 0
   let d = math.random(4)
   for i in countup(0, 3):
      var dx, dy: int
      case (d + i) mod 4
      of 0: dx = 1
      of 1: dx = -1
      of 2: dy = 1
      else: dy = -1
      let
         nx = x + dx
         ny = y + dy
         nx2 = x + 2 * dx
         ny2 = y + 2 * dy
      if maze[ny][nx] == 1 and maze[ny2][nx2] == 1:
         maze[ny][nx] = 0
         carveMaze(maze, nx2, ny2)

proc generateMaze(): MazeT =
   var Presult: ref MazeT
   new(Presult)
   initMaze(Presult)
   carveMaze(Presult, 2, 2)
   Presult[1][2] = 0
   Presult[height - 2][width - 3] = 0
   return Presult[]

showMaze(generateMaze())
